File: tags.py
"""Derived listing labels parsed from free text (title + description).

Pure, read-layer feature: no DB columns, no re-scrape. The app calls `extract(listing)`
on already-stored listings to surface two facts buyers care about but portals bury in the
blurb:

  • facing direction  — "north facing", "facing north-east", "east-facing" -> "North-East"
  • park-facing       — "overlooking the park", "facing a park", "park facing" -> True

Kept deliberately small (regex over text); see CLAUDE.md — do not over-engineer. The 8
compass labels in DIRECTIONS double as the UI filter options.
"""

import re

# Canonical labels, in compass order. Used as the facing-filter options in the UI.
DIRECTIONS = ["North", "South", "East", "West",
              "North-East", "North-West", "South-East", "South-West"]

# Short forms for compact chips/labels (N, NE, ...).
ABBR = {"North": "N", "South": "S", "East": "E", "West": "W",
        "North-East": "NE", "North-West": "NW",
        "South-East": "SE", "South-West": "SW"}

# Compound directions are listed first so the regex alternation prefers them over the
# bare "north"/"east" that they contain.
_DIR = (r"north[\s\-/]?east|north[\s\-/]?west|south[\s\-/]?east|south[\s\-/]?west"
        r"|north|south|east|west")

# A direction sitting next to a "facing" word, in either order:
#   "<dir> facing"  ("north-east facing", "east facing", "west faced")
#   "facing <dir>"  ("facing north east", "facing: south")
_FACING_RE = re.compile(
    rf"(?:({_DIR})[\s\-]*(?:facing|faced|face))|(?:facing[\s:\-]*({_DIR}))", re.I)

# "park facing", or a proximity word ("overlooking / facing / in front of / opposite /
# adjacent to ...") followed within a few words by "park".
_PARK_RE = re.compile(
    r"park[\s\-]?facing"
    r"|(?:overlook(?:s|ing)?|facing|in\s+front\s+of|front\s+of|opposite"
    r"|adjoining|adjacent\s+to|abut(?:s|ting)?|next\s+to|besides?)"
    r"\s+(?:the\s+|a\s+)?(?:[a-z0-9]+\s+){0,3}park",
    re.I)

_COMPOUND = {"northeast": "North-East", "northwest": "North-West",
             "southeast": "South-East", "southwest": "South-West"}
_SIMPLE = {"north": "North", "south": "South", "east": "East", "west": "West"}


def _canon(token: str) -> str | None:
    t = re.sub(r"[\s\-/]", "", token.lower())
    return _COMPOUND.get(t) or _SIMPLE.get(t)


def facing(text: str | None) -> str | None:
    """First compass direction described as the property's facing, or None.
    Returns one of DIRECTIONS, e.g. "North-East"."""
    if not text:
        return None
    m = _FACING_RE.search(text)
    if not m:
        return None
    return _canon(m.group(1) or m.group(2))


def park_facing(text: str | None) -> bool:
    """True if the text says the property faces / overlooks / fronts a park."""
    return bool(text and _PARK_RE.search(text))


def extract(listing: dict) -> dict:
    """Derived labels for one listing dict. Reads title + description.
    Returns {"facing": <label|None>, "park": <bool>}."""
    text = " ".join(str(listing.get(k) or "") for k in ("title", "description"))
    return {"facing": facing(text), "park": park_facing(text)}

File: apps/web/src/lib/tags.ts
// Derived listing labels parsed from free text (title + description).
//
// Read-layer feature: no DB columns, no re-scrape. Surfaces two facts buyers scan for
// but portals bury in the blurb:
//   • facing direction  — "north facing", "facing north-east", "east-facing" -> "North-East"
//   • park-facing       — "overlooking the park", "facing a park", "park facing" -> true
//
// Mirror of the Python tags.py on the v1/Streamlit side. Kept deliberately small.

import type { Match } from "@/lib/types";

// Canonical labels in compass order — also the facing-filter options in the UI.
export const DIRECTIONS = [
  "North", "South", "East", "West",
  "North-East", "North-West", "South-East", "South-West",
] as const;
export type Direction = (typeof DIRECTIONS)[number];

// Short forms for compact chips/labels (N, NE, …).
export const DIR_ABBR: Record<Direction, string> = {
  "North": "N", "South": "S", "East": "E", "West": "W",
  "North-East": "NE", "North-West": "NW", "South-East": "SE", "South-West": "SW",
};

// Compound directions first so the alternation prefers them over the bare words inside.
const DIR = "north[\\s\\-/]?east|north[\\s\\-/]?west|south[\\s\\-/]?east|south[\\s\\-/]?west|north|south|east|west";

// "<dir> facing" or "facing <dir>".
const FACING_RE = new RegExp(`(?:(${DIR})[\\s\\-]*(?:facing|faced|face))|(?:facing[\\s:\\-]*(${DIR}))`, "i");

// "park facing", or a proximity word ("overlooking / facing / in front of / opposite /
// adjacent to …") followed within a few words by "park".
const PARK_RE = new RegExp(
  "park[\\s\\-]?facing" +
  "|(?:overlook(?:s|ing)?|facing|in\\s+front\\s+of|front\\s+of|opposite" +
  "|adjoining|adjacent\\s+to|abut(?:s|ting)?|next\\s+to|besides?)" +
  "\\s+(?:the\\s+|a\\s+)?(?:[a-z0-9]+\\s+){0,3}park",
  "i");

const COMPOUND: Record<string, Direction> = {
  northeast: "North-East", northwest: "North-West",
  southeast: "South-East", southwest: "South-West",
};
const SIMPLE: Record<string, Direction> = {
  north: "North", south: "South", east: "East", west: "West",
};

function canon(token: string): Direction | null {
  const t = token.toLowerCase().replace(/[\s\-/]/g, "");
  return COMPOUND[t] ?? SIMPLE[t] ?? null;
}

/** First compass direction described as the property's facing, or null. */
export function facing(text?: string | null): Direction | null {
  if (!text) return null;
  const m = FACING_RE.exec(text);
  if (!m) return null;
  return canon(m[1] || m[2]);
}

/** True if the text says the property faces / overlooks / fronts a park. */
export function parkFacing(text?: string | null): boolean {
  return !!text && PARK_RE.test(text);
}

const WORD_FLOORS: Record<string, number> = { single: 1, double: 2, triple: 3, four: 4 };

/** Floor layout stated in the text — "G+2", "3 floors", "double storey" — or null. */
export function floors(text?: string | null): string | null {
  if (!text) return null;
  let m = /\b(?:g|ground)\s*\+\s*(\d{1,2})\b/i.exec(text);          // "G+2" / "ground + 2"
  if (m) return `G+${m[1]}`;
  m = /\b(\d{1,2})\s*(?:storey|story|storied|floors?)\b/i.exec(text);  // "3 floors"
  if (m) return `${m[1]} floor${m[1] === "1" ? "" : "s"}`;
  m = /\b(single|double|triple|four)[\s-]*(?:storey|story|storied|floors?)\b/i.exec(text);
  if (m) { const n = WORD_FLOORS[m[1].toLowerCase()]; return `${n} floor${n === 1 ? "" : "s"}`; }
  return null;
}

const ROAD_UNIT = "m|mtr|mtrs|meter|meters|metre|metres|ft|feet";
// "<n> <unit> (wide) road" or "road (width/of) <n> <unit>" → normalized "30m road".
const ROAD_RE_A = new RegExp(
  `\\b(\\d{1,3}(?:\\.\\d+)?)\\s*(${ROAD_UNIT})\\b\\.?\\s*(?:wide\\s+)?(?:mtr\\.?\\s*)?road`, "i");
const ROAD_RE_B = new RegExp(
  `\\broad\\s*(?:width|of|is)?\\s*[:\\-]?\\s*(\\d{1,3}(?:\\.\\d+)?)\\s*(${ROAD_UNIT})\\b`, "i");

/** Approach-road width stated in the text — "30m road", "60ft road" — or null. */
export function roadWidth(text?: string | null): string | null {
  if (!text) return null;
  const m = ROAD_RE_A.exec(text) || ROAD_RE_B.exec(text);
  if (!m) return null;
  return `${m[1]}${/^f/i.test(m[2]) ? "ft" : "m"} road`;
}

// Boolean amenities parsed from the text. `key` is also the PropertyFeatures field and the
// URL filter token; the chip/checkbox show `icon` + `label`. Order = chip/filter order.
export const AMENITIES = [
  { key: "park", icon: "🌳", label: "Park", re: PARK_RE },
  { key: "basement", icon: "🔻", label: "Basement", re: /\bbasement\b/i },
  { key: "stilt", icon: "🚗", label: "Stilt", re: /\bstilt\b/i },
  { key: "corner", icon: "📐", label: "Corner", re: /\bcorner\b/i },
  { key: "lift", icon: "🛗", label: "Lift", re: /\b(?:lift|elevator)\b/i },
] as const;
export type AmenityKey = (typeof AMENITIES)[number]["key"];

export type PropertyFeatures = {
  facing: Direction | null;
  floors: string | null;
  road: string | null;
} & Record<AmenityKey, boolean>;

/** Derived labels for one match, parsed from its title + description. */
export function featureTags(m: Pick<Match, "title" | "description">): PropertyFeatures {
  const text = `${m.title ?? ""} ${m.description ?? ""}`;
  const out = { facing: facing(text), floors: floors(text), road: roadWidth(text) } as PropertyFeatures;
  for (const a of AMENITIES) out[a.key] = a.re.test(text);
  return out;
}

/** Whether a match has any derived feature label (for conditionally rendering chips). */
export function hasFeatures(m: Pick<Match, "title" | "description">): boolean {
  const t = featureTags(m);
  return !!t.facing || !!t.floors || !!t.road || AMENITIES.some((a) => t[a.key]);
}
